plat/common/memory.c

```diff
@@ -137,3 +139,217 @@ void *ukplat_memregion_alloc(__sz size, int type)
+	case UKPLAT_MEMRT_DEVICETREE:
+	case UKPLAT_MEMRT_KERNEL:
+		return 1;
+	case UKPLAT_MEMRT_RESERVED:
+		return 2;
+	default:
+		return -1;
+	}
+}
+
+/* Memory region with lower priority must be adjusted in favor of the one
+ * with higher priority, e.g. if left memory region is of lower priority but
+ * contains the right memory region of higher priority, then split the left one
+ * in two, by adjusting the current left one and inserting a new memory region
+ * descriptor.
+ */
+static inline void overlapping_mrd_fixup(struct ukplat_memregion_list *list,
+					 struct ukplat_memregion_desc *const ml,
+					 struct ukplat_memregion_desc *const mr,
+					 int ml_prio, int mr_prio,
+					 __u32 lidx __unused, __u32 ridx)
+{
+	/* If left memory region is of higher priority */
+	if (ml_prio > mr_prio) {
+		/* If the right region is contained within the left region,
+		 * drop it entirely
+		 */
+		if (RANGE_CONTAIN(ml->pbase, ml->len, mr->pbase, mr->len)) {
+			mr->len = 0;
+
+		/* If the right region has a part of itself in the left region,
+		 * drop that part of the right region only
+		 */
+		} else {
+			mr->len -= ml->pbase + ml->len - mr->pbase;
+			mr->pbase = ml->pbase + ml->len;
+			mr->vbase = mr->pbase;
+		}
+
+	/* If left memory region is of lower priority */
+	} else {
+		/* If the left memory region is contained within the right
+		 * region, drop it entirely
+		 */
+		if (RANGE_CONTAIN(mr->pbase, mr->len, ml->pbase, ml->len)) {
+			ml->len = 0;
+
+		/* If the left region has a part of itself in the right region,
+		 * drop that part of the left region only and split by creating
+		 * a new one if the left region is larger than the right region.
+		 */
+		} else {
+			if (RANGE_CONTAIN(ml->pbase, ml->len,
+					  mr->pbase, mr->len))
+				/* Ignore insertion failure as there is nothing
+				 * we can do about it and it is not worth caring
+				 * about.
+				 */
+				ukplat_memregion_list_insert_at_idx(list,
+					&(struct ukplat_memregion_desc){
+						.vbase = mr->pbase + mr->len,
+						.pbase = mr->pbase + mr->len,
+						.len   = ml->pbase + ml->len -
+							 mr->pbase - mr->len,
+						.type  = ml->type,
+						.flags = ml->flags
+					}, ridx + 1);
+
+			ml->len = mr->pbase - ml->pbase;
+		}
+	}
+}
+
+int ukplat_memregion_list_coalesce(struct ukplat_memregion_list *list)
+{
+	struct ukplat_memregion_desc *m, *ml, *mr;
+	int ml_prio, mr_prio;
+	__u32 i;
+
+	i = 0;
+	m = list->mrds;
+	while (i + 1 < list->count) {
+		/* Make sure first that they are ordered. If not, swap them */
+		if (m[i].pbase > m[i + 1].pbase ||
+		    (m[i].pbase == m[i + 1].pbase &&
+		     m[i].pbase + m[i].len > m[i + 1].pbase + m[i + 1].len)) {
+			struct ukplat_memregion_desc tmp;
+
+			tmp = m[i];
+			m[i] = m[i + 1];
+			m[i + 1] = tmp;
+		}
+		ml = &m[i];
+		mr = &m[i + 1];
+		ml_prio = get_mrd_prio(ml);
+		mr_prio = get_mrd_prio(mr);
+
+		/* If they overlap */
+		if (RANGE_OVERLAP(ml->pbase,  ml->len, mr->pbase, mr->len)) {
```

Comment:
I would solve this by allocating at least page-sized regions, as you have said. But, as you have also mentioned, you end up having to enforce this in the insertion API's as well, meaning that the previous boot phase that passes memory map related stuff will have to ensure that the raw received memory regions do not exist in the same mapping.

This is the solution that I would come up with now: a separate function, called something among the lines of `ukplat_memregion_list_page_overlap_adjust`, that should be called after `ukplat_memregion_list_coalesce`, that simply iterates through memory regions and removes `UKPLAT_MEMRF_MAP` if an in-page overlap is found. I would definetely not do this in the coalescing function as it would get out of its signature's scope.

So that is one suggestion. However, I still do not like it, as I believe that `ukplat_memregion_*` API's should have nothing to do with virtual memory or care about it as that kind of beats the purpose of having a paging API and a memregion API. I believe that `ukplat_memregion_*` API should only do what their names imply and take care of maintaining a proper raw memory map representation. Therefore, I think this case should be taken care of by `ukplat_paging_init`'s loop that maps memory regions with `UKPLAT_MEMRF_MAP`. Especially since it also aligns them to page-sized regions as well, making it already very easy to keep track of which page was mapped previously as they are already sorted ascendingly.

Let me know what you think of these alternate solutions. I would go with the latter.

Reply:
> I would solve this by allocating at least page-sized regions

Just to be clear: multiple of mapping granularity size (4K, 2M, 1G) , not at least page-sized

> Let me know what you think of these alternate solutions. I would go with the latter.

I understand the argumentation towards separation, but in practice I don't think you can ignore PT requirements until you reach the point you map, otherwise you could potentially end up with regions with conflicting attributes. 
Consider for example the bug that triggered this discussion in the first place, where the `cmdl` and `dtb` regions were in the same page. The paging API could merge these into a single mapping if they had consistent protections and other PT attributes. But in any other case, you would end up with `-EEXIST`, which is what we are trying to avoid here. Because of that I see no choice other than enforcing regions to be of mapping granularity size when they are created. Unless I missed something in your proposal, that is.

Reply:
> otherwise you could potentially end up with regions with conflicting attributes.

Ah, very good point, yes! Right, I am only going to make sure this does not happen with `ukplat_memregion_alloc` (allocate page-aligned in multiple of page-sized) only though. I will leave the rest the same and only let this function keep this condition in mind (should also add a comment for this). This means I will also remove the code snippet that takes care of "in-image memory holes".